### PR TITLE
Add environments support to SDK with SDK key initialization

### PR DESCRIPTION
**Pull request by mikeproeng37**

### CODE REVIEW

File: OptimizelySDKShared/OptimizelySDKShared/NSString+OPTLYCategory.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
+ * See the License for the specific language governing permissions and      *
+ * limitations under the License.                                           *
+ ***************************************************************************/
+
+
+#import <Foundation/Foundation.h>
+
+@interface NSString(OPTLYCategory)
+- (BOOL)isValidKeyString;
```
[the remark this answers is not in the document]
Comment (kellyroach-optimizely) [reply OptimizelySDKShared/OptimizelySDKShared/NSString+OPTLYCategory.h]: COMPILING OptimizelySDKiOSUniversal TARGET GETS "warning: umbrella header"

```
<module-includes>:1:1: warning: umbrella header for module 'OptimizelySDKiOS' does not include header '/Users/kroach/objective-c-sdk-environments/OptimizelySDKShared/OptimizelySDKShared/NSString+OPTLYCategory.h' [-Wincomplete-umbrella]
#import "Headers/OptimizelySDKiOS.h"
^
1 warning generated.
1 warning generated.
```

NOTE: "... NSString+OPTLYCategory.h' [-Wincomplete-umbrella]"

**Review by kellyroach-optimizely — CHANGES_REQUESTED** [review]

**Comment by mikeproeng37** [comment]
Changes look good to me @kellyroach-optimizely

**Comment by kellyroach-optimizely** [comment]
TEST:
* sh ./Scripts/build_all.sh PASSED
* 4 Analyze *.framework's PASSED
* 2 Analyze demo apps PASSED
* 8+1 iOS test suites PASSED (*)
* 7+1 tvOS test suites PASSED
* 2 demo apps run PASSED
* 2 demo apps predicted results on OPTIMIZELY.COM PASSED
```
               iOS    tvOS
variation_a    0/7    0/16
variation_b    9/9    11/11
```
* 2 demo apps "iTunes Connect" "Validate..." PASSED

**Comment by kellyroach-optimizely** [comment]
TEST AGAIN
Due to the OPTLY_DATAFILE_URL commit.
* sh ./Scripts/build_all.sh PASSED
* 4 Analyze *.framework's PASSED
* 2 Analyze demo apps PASSED
* 8+1 iOS test suites PASSED
* iOS demo app (8415331295) predicted results on OPTIMIZELY.COM PASSED
```
               iOS
variation_a    0/14
variation_b    21/21
```

**Comment by kellyroach-optimizely** [comment]
CODE REVIEW
Comparison with the referenced SDK android-sdk .
We'd summarize the objective-c-sdk changes by saying:
* In the new world, you can pass either of
    * projectId (the old way)
    * sdkKey (the new way)
to yield a "datafileConfig" internally which is passed around in the
places in the SDK which used to expect a mere "projectId" .  The
"datafileConfig" now plays this role.
* A valid datafileConfig is a (projectId,sdkKey) pair where
exactly one of these strings is non-nil valid string (isValidKeyString:).
That means not nil, not empty, and not containing space chars.

Two deviations of objective-c-sdk from the android-sdk were noted:
* Both SDK's are replacing internal use of "projectId" by something.
The objective-c-sdk is using datafileConfig .  The android-sdk is using
datafileConfigKey .  In an OOP, we think datafileConfig is preferable
and android-sdk could be slightly improved.  However, there isn't
that much difference.
* The objective-c-sdk objective-c-sdk is pickier about the
isValidKeyString: requiring
``((s != nil) && ![s isEqualToString:@""] && ![s containsString:@" "])``
whereas android-sdk is only checking for
``(s != null)``
We communicated this to Mike Ng.  We're told android-sdk should likely
be doing the same check as objective-c-sdk .

**Comment by kellyroach-optimizely** [comment]
CONCLUDE: APPROVING
We have:
* Mike Ng's "Changes look good to me @kellyroach-optimizely"
* TEST PASSED
* CODE REVIEW PASSED

**Review by kellyroach-optimizely — APPROVED** [review]

**Pull request merged**